File: backend/src/services/configurator/pricingEngine.js
'use strict';

/**
 * pricingEngine.js — pure functions, no I/O, no DB.
 *
 * Direct port of `config/backend/app/services/quotation_calc.py`
 * (CALC_VERSION 1.1.0). Numerical results MUST match the Python
 * implementation operator-for-operator so the new Forge backend
 * produces identical quotes to the legacy configurator.
 *
 * Inputs / outputs mirror the Pydantic models in
 * `config/backend/app/schemas/quotation_calc.py` (see LABOR_CATEGORIES).
 */

const CALC_VERSION = '1.1.0';

/** Order matters — matches Python LABOR_CATEGORIES list. */
const LABOR_CATEGORIES = ['CU', 'ASM', 'CNT', 'QC', 'TST', 'ENG', 'CAD'];

// ── Helpers ────────────────────────────────────────────────────────────────

/**
 * Excel-style ROUNDUP. Mirrors the Python helper exactly:
 *   factor < 0  → round UP to the nearest 10**abs(factor)
 *   factor >= 0 → ceil(value * 10**factor) / 10**factor   (decimal shift)
 */
function roundup(value, factor) {
  if (factor < 0) {
    const base = Math.pow(10, Math.abs(factor));
    return Math.ceil(value / base) * base;
  }
  const base = Math.pow(10, factor);
  return Math.ceil(value * base) / base;
}

/**
 * Add `businessDays` skipping weekends and the optional holidays Set.
 * Matches the Python `business_day_add` helper.
 *
 * @param {Date} start
 * @param {number} businessDays
 * @param {Set<string>} [holidaySet]  ISO `YYYY-MM-DD` strings
 * @returns {Date}
 */
function businessDayAdd(start, businessDays, holidaySet) {
  if (!businessDays || businessDays <= 0) return new Date(start.getTime());
  const holidays = holidaySet || new Set();
  const current = new Date(start.getTime());
  let added = 0;
  while (added < businessDays) {
    current.setUTCDate(current.getUTCDate() + 1);
    const dow = current.getUTCDay(); // 0=Sun .. 6=Sat
    if (dow === 0 || dow === 6) continue;
    const iso = current.toISOString().slice(0, 10);
    if (holidays.has(iso)) continue;
    added += 1;
  }
  return current;
}

const num = (v, fallback = 0) => (typeof v === 'number' && Number.isFinite(v) ? v : fallback);

// ── Sub-engines (pure — exposed for testability) ───────────────────────────

/**
 * Compute the per-section material/labour/copper breakdown.
 * Mirrors the per-section loop inside Python `compute_quote`.
 */
function computeSectionBreakdown(sections, lookup) {
  const sectionBreakdowns = [];
  const labourCosts = Object.fromEntries(LABOR_CATEGORIES.map((c) => [c, 0]));
  const labourHours = Object.fromEntries(LABOR_CATEGORIES.map((c) => [c, 0]));
  let materialTotal = 0;
  let sectionCostTotal = 0;
  let copperTotal = 0;

  for (const sec of sections) {
    const qty = num(sec.qty, 0);
    const unitMaterial = num(sec.unit_material_cost, 0);
    const matTotal = unitMaterial * qty;
    materialTotal += matTotal;

    const copper = num(sec.copper_weight_per_unit, 0) * qty;
    copperTotal += copper;

    const perCat = {};
    let labourSumCost = 0;
    const hpu = sec.labor_hours_per_unit || {};
    for (const cat of LABOR_CATEGORIES) {
      const hoursPerUnit = num(hpu[cat], 0);
      const hours = hoursPerUnit * qty;
      const rate = num(lookup[`LBR_${cat}_rate`], 0);
      const cost = hours * rate;
      perCat[cat] = { hours, cost, rate };
      labourCosts[cat] += cost;
      labourHours[cat] += hours;
      labourSumCost += cost;
    }

    const sectionTotal = matTotal + labourSumCost;
    sectionCostTotal += sectionTotal;

    sectionBreakdowns.push({
      id: sec.id,
      description: sec.description || '',
      qty,
      material_total: matTotal,
      labor: perCat,
      section_total: sectionTotal,
      copper_total: copper,
    });
  }

  return {
    section_breakdown: sectionBreakdowns,
    labor_costs: labourCosts,
    labor_hours: labourHours,
    material_total: materialTotal,
    section_cost_total: sectionCostTotal,
    copper_total: copperTotal,
  };
}

/**
 * Group cross-section line-item adders by `desc`.
 * Mirrors the Python adders_grouped + total_line_adders block.
 */
function aggregateAdders(sections) {
  const map = new Map();
  for (const sec of sections) {
    for (const li of sec.line_items || []) {
      const desc = String(li.desc || '');
      map.set(desc, num(map.get(desc), 0) + num(li.value, 0));
    }
  }
  // sort keys ascending — Python uses sorted(adders_map.items())
  const grouped = Array.from(map.entries())
    .sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0))
    .map(([desc, total]) => ({ desc, total }));
  const totalLineAdders = grouped.reduce((acc, a) => acc + a.total, 0);
  return { adders_grouped: grouped, total_line_adders: totalLineAdders };
}

/**
 * Compute target / rounded price + actual gross-margin.
 * Mirrors the pricing block at the bottom of Python `compute_quote`.
 */
function computePricing({ totalCost, pricing }) {
  const strategy = pricing.strategy || 'DESIRED GM%';
  let targetPrice;
  if (strategy === 'DESIRED GM%') {
    const gm = num(pricing.desired_gm_pct, 0);
    if (gm >= 1) {
      throw new Error('desired_gm_pct must be < 1 (i.e. 0.30 not 30)');
    }
    targetPrice = totalCost / (1 - gm);
  } else {
    targetPrice = num(pricing.desired_price, totalCost);
  }
  const roundupFactor = Number.isInteger(pricing.roundup_factor)
    ? pricing.roundup_factor
    : -1;
  const roundedPrice = roundup(targetPrice, roundupFactor);
  const actualProfit = roundedPrice - totalCost;
  const actualGm = roundedPrice ? actualProfit / roundedPrice : 0;
  return {
    target_price: targetPrice,
    rounded_price: roundedPrice,
    actual_profit: actualProfit,
    actual_gm: actualGm,
    roundup_factor: roundupFactor,
  };
}

/**
 * Compute the dependent schedule dates.
 * Mirrors the schedule block of Python `compute_quote`.
 */
function computeSchedule(schedule, holidays = []) {
  const now = new Date();
  const orderDate = schedule.order_date
    ? new Date(schedule.order_date)
    : new Date(now.getTime() + 15 * 24 * 60 * 60 * 1000);
  const holidaySet = new Set((holidays || []).map((d) => String(d).slice(0, 10)));

  const weeksToBd = (weeks) => weeks * 5;

  const longLeadSubDate = schedule.long_lead_sub_weeks
    ? businessDayAdd(orderDate, weeksToBd(schedule.long_lead_sub_weeks), holidaySet)
    : null;
  const engSubDate = schedule.eng_sub_weeks
    ? businessDayAdd(orderDate, weeksToBd(schedule.eng_sub_weeks), holidaySet)
    : null;
  const releaseDate = engSubDate && schedule.sub_approve_weeks
    ? businessDayAdd(engSubDate, weeksToBd(schedule.sub_approve_weeks), holidaySet)
    : engSubDate;

  const totalRtsWeeks = num(schedule.lead_time_weeks, 0) + num(schedule.mfg_time_weeks, 0);
  const rtsDate = releaseDate && totalRtsWeeks
    ? businessDayAdd(releaseDate, weeksToBd(totalRtsWeeks), holidaySet)
    : releaseDate;

  const longLeadApproveDate = schedule.long_lead_approve_weeks
    ? businessDayAdd(orderDate, weeksToBd(schedule.long_lead_approve_weeks), holidaySet)
    : null;

  return {
    order_date: orderDate.toISOString(),
    long_lead_sub_date: longLeadSubDate ? longLeadSubDate.toISOString() : null,
    long_lead_approve_date: longLeadApproveDate ? longLeadApproveDate.toISOString() : null,
    eng_sub_date: engSubDate ? engSubDate.toISOString() : null,
    release_date: releaseDate ? releaseDate.toISOString() : null,
    rts_date: rtsDate ? rtsDate.toISOString() : null,
  };
}

// ── Top-level entry ────────────────────────────────────────────────────────

/**
 * Compute a full quote — direct port of Python `compute_quote(QuoteInput)`.
 *
 * Input shape (verbatim from Pydantic schema):
 *   {
 *     sections: SectionInput[],
 *     lookup:   LookupRates,
 *     pricing:  PricingStrategy,
 *     schedule: ScheduleInput,
 *     holidays?: string[]   // ISO YYYY-MM-DD
 *   }
 *
 * Output shape mirrors `QuoteComputationResult`.
 *
 * Determinism contract: given identical inputs, output MUST equal the
 * Python implementation to the float-precision limit. No rounding is
 * introduced beyond the final `roundup(targetPrice, …)` call.
 */
function computeQuote(data) {
  if (!data || !Array.isArray(data.sections)) {
    throw new Error('computeQuote: data.sections must be an array');
  }
  if (!data.lookup) throw new Error('computeQuote: data.lookup is required');
  if (!data.pricing) throw new Error('computeQuote: data.pricing is required');

  const generatedAt = new Date().toISOString();

  const breakdown = computeSectionBreakdown(data.sections, data.lookup);
  const adders = aggregateAdders(data.sections);

  // Overhead / copper enhancements (optional, piggyback on lookup)
  const overheadPct = num(data.lookup.OVERHEAD_PCT, 0);
  const copperRate = num(data.lookup.COPPER_RATE_PER_LB, 0);
  const copperCost = breakdown.copper_total * copperRate;
  const baseCost = breakdown.section_cost_total + adders.total_line_adders;
  const overheadAmount = baseCost * overheadPct;
  const totalCost = baseCost + overheadAmount + copperCost;

  const pricing = computePricing({ totalCost, pricing: data.pricing });
  const schedule = computeSchedule(data.schedule || {}, data.holidays || []);

  return {
    generated_at: generatedAt,
    calc_version: CALC_VERSION,
    section_breakdown: breakdown.section_breakdown,
    totals: {
      material_total: breakdown.material_total,
      section_cost_total: breakdown.section_cost_total,
      overhead_amount: overheadAmount,
      copper_cost: copperCost,
    },
    labor_costs: breakdown.labor_costs,
    labor_hours: breakdown.labor_hours,
    adders_grouped: adders.adders_grouped,
    total_line_adders: adders.total_line_adders,
    total_cost: totalCost,
    pricing,
    schedule,
    copper_total: breakdown.copper_total,
  };
}

module.exports = {
  CALC_VERSION,
  LABOR_CATEGORIES,
  roundup,
  businessDayAdd,
  computeSectionBreakdown,
  aggregateAdders,
  computePricing,
  computeSchedule,
  computeQuote,
};

File: java-backend/src/main/java/com/forge/configurator/engine/PricingEngine.java
package com.forge.configurator.engine;

import java.time.Instant;
import java.time.LocalDate;
import java.time.ZoneOffset;
import java.time.temporal.ChronoUnit;
import java.util.*;

public class PricingEngine {
    public static final String CALC_VERSION = "1.1.0";
    public static final List<String> LABOR_CATEGORIES = List.of("CU", "ASM", "CNT", "QC", "TST", "ENG", "CAD");

    public record SectionLineItem(String desc, double value) {
    }

    public record SectionInput(String id,
                               String description,
                               double unitMaterialCost,
                               int qty,
                               Map<String, Double> laborHoursPerUnit,
                               double copperWeightPerUnit,
                               List<SectionLineItem> lineItems) {
    }

    public record PricingStrategy(String strategy, Double desiredGmPct, Double desiredPrice, int roundupFactor) {
    }

    public record ScheduleInput(Instant orderDate,
                                int longLeadSubWeeks,
                                int longLeadApproveWeeks,
                                int engSubWeeks,
                                int subApproveWeeks,
                                int leadTimeWeeks,
                                int mfgTimeWeeks) {
    }

    public record AddersGroupedItem(String desc, double total) {
    }

    public record LaborCategoryBreakdown(double hours, double cost, double rate) {
    }

    public record SectionBreakdown(String id,
                                   String description,
                                   int qty,
                                   double materialTotal,
                                   Map<String, LaborCategoryBreakdown> labor,
                                   double sectionTotal,
                                   double copperTotal) {
    }

    public record PricingResult(double targetPrice,
                                double roundedPrice,
                                double actualProfit,
                                double actualGm,
                                int roundupFactor) {
    }

    public record ScheduleResult(Instant orderDate,
                                 Instant longLeadSubDate,
                                 Instant longLeadApproveDate,
                                 Instant engSubDate,
                                 Instant releaseDate,
                                 Instant rtsDate) {
    }

    public record QuoteResult(Instant generatedAt,
                              String calcVersion,
                              List<SectionBreakdown> sectionBreakdown,
                              Map<String, Double> totals,
                              Map<String, Double> laborCosts,
                              Map<String, Double> laborHours,
                              List<AddersGroupedItem> addersGrouped,
                              double totalLineAdders,
                              double totalCost,
                              PricingResult pricing,
                              ScheduleResult schedule,
                              double copperTotal) {
    }

    public record QuoteInput(List<SectionInput> sections,
                             Map<String, Double> lookup,
                             PricingStrategy pricing,
                             ScheduleInput schedule,
                             List<LocalDate> holidays) {
    }

    public static double roundup(double value, int factor) {
        if (factor < 0) {
            double base = Math.pow(10, Math.abs(factor));
            return Math.ceil(value / base) * base;
        }
        double base = Math.pow(10, factor);
        return Math.ceil(value * base) / base;
    }

    private static Instant businessDayAdd(Instant start, int businessDays, Set<LocalDate> holidays) {
        if (businessDays <= 0) {
            return start;
        }
        LocalDate date = start.atOffset(ZoneOffset.UTC).toLocalDate();
        int added = 0;
        while (added < businessDays) {
            date = date.plusDays(1);
            if (date.getDayOfWeek().getValue() >= 6) {
                continue;
            }
            if (holidays.contains(date)) {
                continue;
            }
            added++;
        }
        return date.atStartOfDay().toInstant(ZoneOffset.UTC);
    }

    public QuoteResult computeQuote(QuoteInput input) {
        if (input.sections() == null) {
            throw new IllegalArgumentException("sections is required");
        }
        if (input.lookup() == null) {
            throw new IllegalArgumentException("lookup is required");
        }
        if (input.pricing() == null) {
            throw new IllegalArgumentException("pricing is required");
        }

        List<SectionBreakdown> sectionBreakdowns = new ArrayList<>();
        Map<String, Double> laborCosts = new LinkedHashMap<>();
        Map<String, Double> laborHours = new LinkedHashMap<>();
        for (String cat : LABOR_CATEGORIES) {
            laborCosts.put(cat, 0.0);
            laborHours.put(cat, 0.0);
        }

        double materialTotal = 0.0;
        double sectionCostTotal = 0.0;
        double copperTotal = 0.0;

        for (SectionInput sec : input.sections()) {
            int qty = Math.max(0, sec.qty());
            double matTotal = sec.unitMaterialCost() * qty;
            materialTotal += matTotal;

            double copper = sec.copperWeightPerUnit() * qty;
            copperTotal += copper;

            Map<String, LaborCategoryBreakdown> perCat = new LinkedHashMap<>();
            double laborSumCost = 0.0;
            for (String cat : LABOR_CATEGORIES) {
                double hpu = num(sec.laborHoursPerUnit().get(cat));
                double hours = hpu * qty;
                double rate = num(input.lookup().get("LBR_" + cat + "_rate"));
                double cost = hours * rate;
                perCat.put(cat, new LaborCategoryBreakdown(hours, cost, rate));

                laborCosts.put(cat, laborCosts.get(cat) + cost);
                laborHours.put(cat, laborHours.get(cat) + hours);
                laborSumCost += cost;
            }

            double sectionTotal = matTotal + laborSumCost;
            sectionCostTotal += sectionTotal;
            sectionBreakdowns.add(new SectionBreakdown(
                    sec.id(),
                    sec.description(),
                    qty,
                    matTotal,
                    perCat,
                    sectionTotal,
                    copper
            ));
        }

        Map<String, Double> addersMap = new HashMap<>();
        for (SectionInput sec : input.sections()) {
            for (SectionLineItem li : sec.lineItems()) {
                addersMap.merge(li.desc(), li.value(), Double::sum);
            }
        }

        List<AddersGroupedItem> addersGrouped = addersMap.entrySet().stream()
                .sorted(Map.Entry.comparingByKey())
                .map(e -> new AddersGroupedItem(e.getKey(), e.getValue()))
                .toList();

        double totalLineAdders = addersGrouped.stream().mapToDouble(AddersGroupedItem::total).sum();

        double overheadPct = num(input.lookup().get("OVERHEAD_PCT"));
        double copperRate = num(input.lookup().get("COPPER_RATE_PER_LB"));
        double copperCost = copperTotal * copperRate;
        double baseCost = sectionCostTotal + totalLineAdders;
        double overheadAmount = baseCost * overheadPct;
        double totalCost = baseCost + overheadAmount + copperCost;

        PricingResult pricing = computePricing(totalCost, input.pricing());
        ScheduleResult schedule = computeSchedule(input.schedule(), input.holidays());

        Map<String, Double> totals = new LinkedHashMap<>();
        totals.put("material_total", materialTotal);
        totals.put("section_cost_total", sectionCostTotal);
        totals.put("overhead_amount", overheadAmount);
        totals.put("copper_cost", copperCost);

        return new QuoteResult(
                Instant.now(),
                CALC_VERSION,
                sectionBreakdowns,
                totals,
                laborCosts,
                laborHours,
                addersGrouped,
                totalLineAdders,
                totalCost,
                pricing,
                schedule,
                copperTotal
        );
    }

    public PricingResult computePricing(double totalCost, PricingStrategy pricing) {
        String strategy = pricing.strategy() == null ? "DESIRED GM%" : pricing.strategy();
        double targetPrice;
        if ("DESIRED GM%".equals(strategy)) {
            double gm = pricing.desiredGmPct() == null ? 0.0 : pricing.desiredGmPct();
            if (gm >= 1) {
                throw new IllegalArgumentException("desired_gm_pct must be < 1");
            }
            targetPrice = totalCost / (1 - gm);
        } else {
            targetPrice = pricing.desiredPrice() == null ? totalCost : pricing.desiredPrice();
        }

        int roundupFactor = pricing.roundupFactor();
        double roundedPrice = roundup(targetPrice, roundupFactor);
        double actualProfit = roundedPrice - totalCost;
        double actualGm = roundedPrice == 0 ? 0 : actualProfit / roundedPrice;

        return new PricingResult(targetPrice, roundedPrice, actualProfit, actualGm, roundupFactor);
    }

    public ScheduleResult computeSchedule(ScheduleInput schedule, List<LocalDate> holidays) {
        ScheduleInput effective = schedule == null
                ? new ScheduleInput(null, 0, 0, 0, 0, 0, 0)
                : schedule;

        Instant now = Instant.now();
        Instant orderDate = effective.orderDate() == null ? now.plus(15, ChronoUnit.DAYS) : effective.orderDate();

        Set<LocalDate> holidaySet = holidays == null ? Set.of() : new HashSet<>(holidays);
        Instant longLeadSub = effective.longLeadSubWeeks() > 0
                ? businessDayAdd(orderDate, effective.longLeadSubWeeks() * 5, holidaySet)
                : null;
        Instant engSub = effective.engSubWeeks() > 0
                ? businessDayAdd(orderDate, effective.engSubWeeks() * 5, holidaySet)
                : null;
        Instant release = engSub != null && effective.subApproveWeeks() > 0
                ? businessDayAdd(engSub, effective.subApproveWeeks() * 5, holidaySet)
                : engSub;

        int totalRtsWeeks = effective.leadTimeWeeks() + effective.mfgTimeWeeks();
        Instant rts = release != null && totalRtsWeeks > 0
                ? businessDayAdd(release, totalRtsWeeks * 5, holidaySet)
                : release;

        Instant longLeadApprove = effective.longLeadApproveWeeks() > 0
                ? businessDayAdd(orderDate, effective.longLeadApproveWeeks() * 5, holidaySet)
                : null;

        return new ScheduleResult(orderDate, longLeadSub, longLeadApprove, engSub, release, rts);
    }

    private double num(Double value) {
        return value == null ? 0.0 : value;
    }
}
